### ISSUE TITLE
Gradient accumulation drops all but the last micro-batch with shared grad buffers (FSDP size 1)

### ISSUE THREAD

**Issue by jlamypoirier** [ServiceNow/Fast-LLM#550]
_Filed by Claude Opus 4.8 (Claude Code). Diagnosis below is reproduced but the fix is a proposal — flagging for investigation._

## Summary

When grad buffers are **shared across stages** (`num_grad_buffers < num_stages`, e.g. anything that sets `zero_stage >= 2`) **and** there is more than one micro-batch per step (`schedule.depth_first_micro_batches > 1`) **and** the trainer runs with **data-parallel / FSDP size 1** (`_fsdp_dim.size == 1`), gradient accumulation is broken: the persistent grad shard ends up holding only the **last** micro-batch's gradient for every stage that shares a buffer. Stages whose buffer is *not* reused mid-step (e.g. the tied embedding) are unaffected.

Net effect: most parameters receive roughly `1/num_micro_batches` of their true gradient, silently. No error is raised.

## Affected code

`fast_llm/engine/multi_stage/fsdp.py`, `Fsdp.reduce_gradients` (around lines 391–407). The contract is: reduce the buffer, then **copy** into the grad shard on a stage's first flush and **add** on subsequent flushes (`accumulate` is passed in by the scheduler). The sharded branch implements both arms:

```python
if self._fsdp_dim.size > 1:
    ...
    if accumulate:
        triton_add(self._grad_shard, out, self._grad_shard)        # subsequent flush: add
    elif not full_precision_gradients:
        triton_copy(self._grad_buffer_local_shard, self._grad_shard)  # first flush: copy
else:
    triton_copy(self._grad_buffer_local_shard, self._grad_shard)   # <-- always copies, ignores `accumulate`
```

The `else` branch (FSDP size 1) ignores the `accumulate` flag and always overwrites the grad shard.

This is the path the existing comment in `fast_llm/engine/schedule/runner.py:186` half-acknowledges:

```python
# Reset gradients
# TODO: This is incorrect with shared buffers.
#   (still works because only the embedding layer doesn't share buffer)
```

— "still works" only holds for a single micro-batch; under gradient accumulation the shared-buffer stages are clobbered.

## Mechanism

The scheduler (`Schedule._setup_reduce_steps`) inserts a reduce whenever a grad buffer is about to be reused by a different stage, and marks it `reduce_accumulate=True` for every flush of a stage after its first. With shared buffers, each stage is flushed **once per micro-batch** (its buffer is immediately needed by the next stage); `runner._reduce` then calls `reduce_gradients(accumulate=...)` followed by `reset_gradients()`.

- micro-batch 0: stage X backward → flush → grad shard ← X's mb0 grad (copy)
- micro-batch 1: stage X backward → flush (`accumulate=True` intended) → grad shard **overwritten** with X's mb1 grad
- … → grad shard holds only the last micro-batch's grad for stage X.

The `reduce_accumulate=True` intent is correct; the FSDP-size-1 branch just never honors it.

## Why it's usually masked

With dedicated buffers (one per stage, the default when ZeRO is off), each stage is flushed exactly **once** — at the end, after autograd has summed all micro-batches into its buffer — so a single copy is the right answer and accumulation is never needed. The bug only surfaces when buffer sharing forces multiple flushes of the same stage within a step.

## Impact

Quietly suppresses the gradient of most parameters by ~`num_micro_batches×` whenever a run uses shared grad buffers + gradient accumulation on an unsharded trainer. Training proceeds without error; the model just learns far more slowly on the affected parameters than the configuration implies.

## Reproduction

One training step, frozen or live weights, on a single trainer process (data-parallel size 1):

1. Configure `schedule.depth_first_micro_batches = N` (N > 1, real gradient accumulation) and force shared grad buffers (e.g. `zero_stage = 2`, i.e. `num_grad_buffers = 2`, with more than 2 stages).
2. Read per-parameter grad L2 after the step.
3. Compare against the same data/weights with `num_grad_buffers >= num_stages` (dedicated buffers), or against a single concatenated forward/backward over the same tokens.

Observed (Qwen2.5-0.5B, 26 stages, 7 micro-batches, frozen weights, one step):

| grad buffers | tied embedding grad L2 | all other params grad L2 |
|---|---|---|
| 3 (shared, `num_grad_buffers=2`) | 169.4 | **31.9** |
| 26 (dedicated, default) | 169.4 | **329.7** |

The tied embedding (its own dedicated buffer) is identical both ways; everything else collapses ~10× under shared buffers. `329.7` matches a single-pass autograd reference, i.e. it is the correct value.

## Proposed fix

Make the FSDP-size-1 branch honor `accumulate`, mirroring the sharded arm:

```python
else:
    if accumulate:
        triton_add(self._grad_shard, self._grad_buffer_local_shard, self._grad_shard)
    else:
        triton_copy(self._grad_buffer_local_shard, self._grad_shard)
```

A regression test would run a multi-micro-batch step under shared buffers (`num_grad_buffers` < num stages, `depth_first_micro_batches > 1`, FSDP size 1) and assert the per-parameter grads equal the dedicated-buffer / single-pass values.

## Scope / caveat

The bug is specific to the `_fsdp_dim.size == 1` branch — i.e. no data-parallel grad sharding. With FSDP size > 1, `reduce_gradients` takes the other branch, which does honor `accumulate` (reduce-scatter into a temp, then `triton_add`), so that configuration should be unaffected. Worth confirming whether the comment at `runner.py:186` should also be revisited once this is fixed.
